### Data Profiling

#### Importing Packages and Data

In [22]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from pathlib import Path

In [23]:
# Setting up Project Paths

project_root = Path.cwd().parent
raw_dir = project_root/"data"/"raw"
csv_files = sorted(raw_dir.rglob("*.csv"))

In [24]:
# Setting up Data Frame

frames = []

for f in csv_files:
    df_part = pd.read_csv(f)
    df_part["source_file"] = f.name
    frames.append(df_part)

df = pd.concat(frames, ignore_index=True)

In [25]:
df

,Crime ID,Month,Reported by,Falls within,Longitude,Latitude,Location,LSOA code,LSOA name,Crime type,Last outcome category,Context,source_file
0,e7b720d0e1302d2d06db7b28b29132eb194864d44d7921...,2024-01,City of London Police,City of London Police,-0.106220,51.518275,On or near B500,E01000916,Camden 027B,Theft from the person,Investigation complete; no suspect identified,NaN,2024-01-city-of-london-street.csv
1,e60a5ac62a80e866453254474137c3206417422c62f0c0...,2024-01,City of London Police,City of London Police,-0.107682,51.517786,On or near B521,E01000917,Camden 027C,Other theft,Investigation complete; no suspect identified,NaN,2024-01-city-of-london-street.csv
2,986f618142ec52b7f254e4b0549da2f17ceeb0e130db6c...,2024-01,City of London Police,City of London Police,-0.111596,51.518281,On or near Chancery Lane,E01000914,Camden 028B,Other theft,Investigation complete; no suspect identified,NaN,2024-01-city-of-london-street.csv
3,05dc27a88748356f6d59b0bd1389710ebfb42b37e565af...,2024-01,City of London Police,City of London Police,-0.111596,51.518281,On or near Chancery Lane,E01000914,Camden 028B,Theft from the person,Status update unavailable,NaN,2024-01-city-of-london-street.csv
4,373d78e2ccec5d05a547cd4bee19045a9e050042a0e6e7...,2024-01,City of London Police,City of London Police,-0.112096,51.515942,On or near Nightclub,E01000914,Camden 028B,Theft from the person,Investigation complete; no suspect identified,NaN,2024-01-city-of-london-street.csv
...,...,...,...,...,...,...,...,...,...,...,...,...,...
2298607,d7bcaa710e61011a6192bd504e86738b100f47b42be464...,2025-12,Metropolitan Police Service,Metropolitan Police Service,-2.171098,52.568065,On or near Warstones Drive,E01010490,Wolverhampton 028B,Violence and sexual offences,Under investigation,NaN,2025-12-metropolitan-street.csv
2298608,60acb42bd7459bb8963f3009bdaf77e8246a1b525a3ce6...,2025-12,Metropolitan Police Service,Metropolitan Police Service,-1.851453,52.110001,On or near Little Road,E01032415,Wychavon 011D,Other crime,Under investigation,NaN,2025-12-metropolitan-street.csv
2298609,1d6177d0d7e23c56d48ea7511406118f0daa205d35c8b7...,2025-12,Metropolitan Police Service,Metropolitan Police Service,-1.870244,52.040086,On or near Station Road,E01032351,Wychavon 018A,Theft from the person,Investigation complete; no suspect identified,NaN,2025-12-metropolitan-street.csv
2298610,86578e92566319f2b8822dd2301965f40a9dfa7e9e6f43...,2025-12,Metropolitan Police Service,Metropolitan Police Service,-2.262334,52.389968,On or near Church Walk,E01032454,Wyre Forest 006D,Violence and sexual offences,Under investigation,NaN,2025-12-metropolitan-street.csv


In [26]:
df.shape

(2298612, 13)

In [27]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 2298612 entries, 0 to 2298611
Data columns (total 13 columns):
 #   Column                 Dtype  
---  ------                 -----  
 0   Crime ID               str    
 1   Month                  str    
 2   Reported by            str    
 3   Falls within           str    
 4   Longitude              float64
 5   Latitude               float64
 6   Location               str    
 7   LSOA code              str    
 8   LSOA name              str    
 9   Crime type             str    
 10  Last outcome category  str    
 11  Context                float64
 12  source_file            str    
dtypes: float64(3), str(10)
memory usage: 751.0 MB


In [28]:
# Missing Values
missing = pd.DataFrame({
    "missing_count": df.isna().sum(),
    "missing_pct": (df.isna().mean() * 100).round(2)
})
print(missing)

                       missing_count  missing_pct
Crime ID                      467898        20.36
Month                              0         0.00
Reported by                        0         0.00
Falls within                       0         0.00
Longitude                       2261         0.10
Latitude                        2261         0.10
Location                           0         0.00
LSOA code                       2262         0.10
LSOA name                       2262         0.10
Crime type                         0         0.00
Last outcome category         467898        20.36
Context                      2298612       100.00
source_file                        0         0.00


In [29]:
# Are all 24 months present, and how many crimes per month?
print(df["Month"].value_counts().sort_index())

# Which Police force reported it in the data?
print(df["Reported by"].value_counts())

# What crime types exist, and are there exactly 14?
print(f"\nCrime types: {df['Crime type'].nunique()}")
print(df["Crime type"].value_counts())

# Are there exact duplicated rows?
dupes = df.duplicated().sum()
print(f"\n Duplicate rows: {dupes:,}")

Month
2024-01     92145
2024-02     91601
2024-03     92348
2024-04     91054
2024-05     98021
2024-06     97581
2024-07    101139
2024-08     98355
2024-09     94728
2024-10    101686
2024-11     98832
2024-12     91831
2025-01     87581
2025-02     84361
2025-03     94334
2025-04     93949
2025-05    100164
2025-06    101615
2025-07    107440
2025-08     99838
2025-09     93969
2025-10     99326
2025-11     94943
2025-12     91771
Name: count, dtype: int64
Reported by
Metropolitan Police Service    2279745
City of London Police            18867
Name: count, dtype: int64

Crime types: 14
Crime type
Violence and sexual offences    527742
Anti-social behaviour           467898
Other theft                     214603
Theft from the person           185684
Vehicle crime                   181114
Shoplifting                     172546
Public order                    114347
Criminal damage and arson       110708
Burglary                        101028
Drugs                            92925
Ro

#### Summary

Dataset: Metropolitan and City of London Police reported recorded crime from January 2024 till December 2025
"Crime ID" and "Last outcome category" have about 20% rows missing due to Anti social behaviour having no offense ID. "Context" is 100% missing and "LSOA" has 2,262 rows missing which is 0.10% so they will be dropped.

Volume: 2,298,612 crimes across 24 months in 48 CSV files (Metroplitan Police and City of London Police).
Average of almost 96,000 crimes per month.

Crime Types: 14 Categories. Top 3: Violence & sexual offences (527,742), Anti-social behaviour (467,898), Other theft (214,603).